File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.*;
import java.util.*;

public class Main {	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st;			
		
		int N = Integer.parseInt(br.readLine());
				
		int[] num = new int[N+6];
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			int d = Integer.parseInt(st.nextToken());
			int s = Integer.parseInt(st.nextToken());
			
			num[i+d] = Math.max(num[i+d], num[i] + s);
			for (int j = i+d; j <= N; j++) {
				num[j] = Math.max(num[j-1], num[j]);
			}
		}
		System.out.println(num[N]);
		br.close();
	}
}

File: 백준/Silver/14501. 퇴사/퇴사.py
import sys
input = sys.stdin.readline

N = int(input())
num = [0 for _ in range(N+6)]
for i in range(N) :
    d,s = map(int, input().split())

    num[i+d] = max(num[i+d], num[i] + s)
    for j in range(i+d+1, N+1) :
        num[j] = max(num[j-1], num[j])

print(num[N])